src/card-draw.ts

```diff
@@ -307,62 +307,79 @@ export function draw(gameData: GameData, configData: ConfigState): Drawing {
     }
   }
 
+  // OK, setup work is done, here's whre we actually draw the cards!
+
+  let redraw = false;
   const drawnCharts: DrawnChart[] = [];
-  /**
-   * Record of how many songs of each bucket index have been drawn so far
-   */
-  const difficultyCounts = new CountingSet<number>();
 
-  // OK, setup work is done, here's whre we actually draw the cards!
-  while (drawnCharts.length < numChartsToRandom) {
-    if (bucketDistribution.length === 0) {
-      // no more songs available to pick in the requested range
-      // will be returning fewer than requested number of charts
-      break;
-    }
+  do {
+    /**
+     * Record of how many songs of each bucket index have been drawn so far
+     */
+    const difficultyCounts = new CountingSet<number>();
 
-    // first pick a difficulty (with priority to minimum draws)
-    let chosenBucketIdx = requiredDrawIndexes.shift();
-    if (chosenBucketIdx === undefined) {
+    while (drawnCharts.length < numChartsToRandom) {
+      if (bucketDistribution.length === 0) {
+        // no more songs available to pick in the requested range
+        // will be returning fewer than requested number of charts
+        break;
+      }
+
+      let chosenBucketIdx = undefined;
       [, chosenBucketIdx] = pickRandomItem(bucketDistribution);
-    }
-    if (chosenBucketIdx === undefined) {
-      // nothing left to draw
-      break;
-    }
-    const selectableCharts = validCharts.get(chosenBucketIdx);
-    if (!selectableCharts) {
-      // something bad happened?!
-      break;
-    }
-    const [randomIndex, randomChart] = pickRandomItem(selectableCharts);
-
-    if (randomChart) {
-      // Save it in our list of drawn charts
-      drawnCharts.push({
-        ...randomChart,
-        // Give this random chart a unique id within this drawing
-        id: `drawn_chart-${nanoid(5)}`,
-        type: CHART_DRAWN,
-      });
-      // remove drawn chart from deck so it cannot be re-drawn
-      selectableCharts.splice(randomIndex, 1);
-      difficultyCounts.add(chosenBucketIdx);
+
+      if (chosenBucketIdx === undefined) {
+        // nothing left to draw
+        break;
+      }
+      const selectableCharts = validCharts.get(chosenBucketIdx);
+      if (!selectableCharts) {
+        // something bad happened?!
+        break;
+      }
+      const [randomIndex, randomChart] = pickRandomItem(selectableCharts);
+
+      if (randomChart) {
+        // Save it in our list of drawn charts
+        drawnCharts.push({
+          ...randomChart,
+          // Give this random chart a unique id within this drawing
+          id: `drawn_chart-${nanoid(5)}`,
+          type: CHART_DRAWN,
+        });
+        // remove drawn chart from deck so it cannot be re-drawn
+        selectableCharts.splice(randomIndex, 1);
```

Comment:
I think the problem with this approach as-is occurs right here, where we remove charts from the "deck" as we draw them. With this looping approach we should be starting with a full deck each iteration, instead of leaving the contents of earlier draws in a "discard pile", so to speak.